folding-schemes/src/folding/hypernova/circuits.rs

```diff
@@ -345,24 +430,320 @@ fn compute_c_gadget<F: PrimeField>(
+            c: vec![C1::ScalarField::one(), C1::ScalarField::one().neg()],
+            M: vec![],
+        };
+        let mut augmented_f_circuit = Self {
+            _c2: PhantomData,
+            _gc2: PhantomData,
+            poseidon_config: poseidon_config.clone(),
+            ccs: initial_ccs.clone(),
+            i: None,
+            i_usize: None,
+            z_0: None,
+            z_i: None,
+            external_inputs: None,
+            u_i_C: None,
+            U_i: None,
+            U_i1_C: None,
+            F: F_circuit,
+            x: None,
+            nimfs_proof: None,
+        };
+        if ccs.is_some() {
+            augmented_f_circuit.ccs = ccs.unwrap();
+        } else {
+            augmented_f_circuit.ccs = augmented_f_circuit.upper_bound_ccs()?;
+        }
+        Ok(augmented_f_circuit)
+    }
+
+    /// This method computes the CCS parameters. This is used because there is a circular
+    /// dependency between the AugmentedFCircuit CCS and the CCS parameters m & n & s & s'.
+    /// For a stable FCircuit circuit, the CCS parameters can be computed in advance and can be
+    /// feed in as parameter for the AugmentedFCircuit::empty method to avoid computing them there.
+    pub fn upper_bound_ccs(&self) -> Result<CCS<C1::ScalarField>, Error> {
+        let mut r1cs = get_r1cs_from_cs::<CF1<C1>>(self.clone()).unwrap();
+        let mut ccs = CCS::from_r1cs(r1cs.clone());
+
+        let z_0 = vec![C1::ScalarField::zero(); self.F.state_len()];
+        let mut W_i =
+            Witness::<C1::ScalarField>::new(vec![C1::ScalarField::zero(); ccs.n - ccs.l - 1]);
+        let mut U_i = LCCCS::<C1>::dummy(ccs.l, ccs.t, ccs.s);
+        let mut w_i = W_i.clone();
+        let mut u_i = CCCS::<C1>::dummy(ccs.l);
+
+        let n_iters = 3;
+
+        for _ in 0..n_iters {
+            let mut transcript_p: PoseidonTranscript<C1> =
+                PoseidonTranscript::<C1>::new(&self.poseidon_config.clone());
+            let (nimfs_proof, U_i1, _) = NIMFS::<C1, PoseidonTranscript<C1>>::prove(
+                &mut transcript_p,
+                &ccs,
+                &[U_i.clone()],
+                &[u_i.clone()],
+                &[W_i.clone()],
+                &[w_i.clone()],
+            )
+            .unwrap();
+
+            let augmented_f_circuit = Self {
+                _c2: PhantomData,
+                _gc2: PhantomData,
+                poseidon_config: self.poseidon_config.clone(),
+                ccs: ccs.clone(),
+                i: Some(C1::ScalarField::zero()),
+                i_usize: Some(0),
+                z_0: Some(z_0.clone()),
+                z_i: Some(z_0.clone()),
+                external_inputs: Some(vec![]),
+                u_i_C: Some(u_i.C),
+                U_i: Some(U_i.clone()),
+                U_i1_C: Some(U_i1.C),
+                F: self.F.clone(),
+                x: Some(C1::ScalarField::zero()),
+                nimfs_proof: Some(nimfs_proof),
+            };
```

Comment:
Also builder patter n can be helpful here too.

Reply:
I think that it is not needed, at least for the moment and foreseable future, the reasoning:

I'm afraid of falling into *premature abstraction* (not only here, but on all the places in which we could add extra methods that would add complexity to the code without a clear advantadge), since this will not be used by the end user of the library, and we only construct this struct in the tests and in the ivc at only two different places.

I've implemented using the build pattern metod locally to see how it would unfold, and I'm not convinced of obtaining any advantadges:

constructing directly the struct as it currently is takes 16 lines (having used rust fmt):
```rust
            let augmented_f_circuit =
                AugmentedFCircuit::<Projective, Projective2, GVar2, CubicFCircuit<Fr>> {
                    poseidon_config: poseidon_config.clone(),
                    ccs: ccs.clone(),
                    i: Some(i),
                    i_usize: Some(i_usize),
                    z_0: Some(z_0.clone()),
                    z_i: Some(z_i.clone()),
                    external_inputs: Some(vec![]),
                    u_i_C: Some(u_i.C),
                    U_i: Some(U_i.clone()),
                    U_i1_C: Some(U_i1.C),
                    F: F_circuit,
                    x: Some(u_i1_x),
                    nimfs_proof: Some(nimfs_proof),
                };
```
(removed the 2 PhantomData lines since they will not exist once the circuit contains the CycleFold logic) 

whereas if we use a builder method like:
```rust
    pub fn build(
        poseidon_config: &PoseidonConfig<CF1<C1>>,
        ccs: CCS<C1::ScalarField>,
        i: C1::ScalarField,
        i_usize: usize,,
        z_0: Vec<C1::ScalarField>,
        z_i: Vec<C1::ScalarField>,
        external_inputs: Vec<C1::ScalarField>,
        U_i: LCCCS<C1>,
        u_i: CCCS<C1>,
        U_i1: LCCCS<C1>,
        F: FC,
        x: Vec<C1::ScalarField>,
        nimfs_proof: NIMFSProof<C1>,
    ) -> Self {
        Self {
            _c2: PhantomData,
            _gc2: PhantomData,
            poseidon_config: poseidon_config.clone(),
            ccs: ccs.clone(),
            i: Some(i),
            i_usize: Some(i_usize),
            z_0: Some(z_0),
            z_i: Some(z_i),
            external_inputs: Some(external_inputs),
            u_i_C: Some(u_i.C),
            U_i: Some(U_i),
            U_i1_C: Some(U_i1.C),
            F,
            x: Some(x[0]),
            nimfs_proof: Some(nimfs_proof),
        }
    }
```

then calling the build method to build the struct will take 15 lines (having used rust fmt):
```rust
            let augmented_f_circuit =
                AugmentedFCircuit::<Projective, Projective2, GVar2, CubicFCircuit<Fr>>::new(
                    poseidon_config,
                    ccs,
                    i,
                    i_usize,
                    z_0,
                    z_i,
                    external_inputs,
                    U_i,
                    u_i,
                    U_i1,
                    F_circuit,
                    x,
                    nimfs_proof,
                );
```

the difference is almost minimal, and it takes the same amount of lines (having used rust fmt), I mean is not a difference from reducing the code, with the extra of adding extra code for the build method which basically constructs the struct.
And also having in mind that this is not used by the user of the library, and only will appear at most 2 times in the ivc proving step, it feels like adding extra code to end up with almost the same code usage to obtain the struct.

How do you see it?